File: strategy/Strategy.cpp
#include <iostream>
#include <string>

using namespace std;

class IOperatorStrategy
{
public:
    virtual void calculate(int a, int b) = 0;
    virtual ~IOperatorStrategy() {}
};

class AddOperator : public IOperatorStrategy
{
public:
    void calculate(int a, int b)
    {
        cout << "The result of " << a << " + " << b << " = " << a + b << endl;
    }
};

class SubOperator : public IOperatorStrategy
{
public:
    void calculate(int a, int b)
    {
        cout << "The result of " << a << " - " << b << " = " << a - b << endl;
    }
};

class MulOperator : public IOperatorStrategy
{
public:
    void calculate(int a, int b)
    {
        cout << "The result of " << a << " * " << b << " = " << a * b << endl;
    }
};

class DivOperator : public IOperatorStrategy
{
public:
    void calculate(int a, int b)
    {
        cout << "The result of " << a << " / " << b << " = " << a / b << endl;
    }
};

class Calculator
{
private:
    IOperatorStrategy *m_strategy;

public:
    Calculator(IOperatorStrategy *strategy) : m_strategy(strategy) {}
    void calculate(int a, int b)
    {
        m_strategy->calculate(a, b);
    }
    void setStrategy(IOperatorStrategy *strategy)
    {
        m_strategy = strategy;
    }
};

// as a client
int main(int argc, char *argv[])
{
    Calculator *calculator = new Calculator(new AddOperator());
    calculator->calculate(1, 2);

    calculator->setStrategy(new SubOperator());
    calculator->calculate(1, 2);

    calculator->setStrategy(new MulOperator());
    calculator->calculate(1, 2);

    calculator->setStrategy(new DivOperator());
    calculator->calculate(1, 2);
    return 0;
}

File: strategy/Strategy.java

public class Strategy {

    private interface IOperatorStrategy {
        public int calculate(int a, int b);
    }

    private class AddOperator implements IOperatorStrategy {
        @Override
        public int calculate(int a, int b) {
            System.out.println("The result of " + a + " + " + b + " =");
            return a + b;
        }
    }

    private class SubOperator implements IOperatorStrategy {
        @Override
        public int calculate(int a, int b) {
            System.out.println("The result of " + a + " - " + b + " =");
            return a - b;
        }
    }

    private class MulOperator implements IOperatorStrategy {
        @Override
        public int calculate(int a, int b) {
            System.out.println("The result of " + a + " * " + b + " =");
            return a * b;
        }
    }

    private class DivOperator implements IOperatorStrategy {
        @Override
        public int calculate(int a, int b) {
            System.out.println("The result of " + a + " / " + b + " =");
            return a / b;
        }
    }

    // Another way to implement the strategy pattern is to use enum and lambda expression.
    private class LambdaStartegy {
        // Each constants will execute the same constructor.
        public enum EnumOperatorStrategy implements IOperatorStrategy {
            ADD((int a, int b)->{
                System.out.println("The result of " + a + " + " + b + " =");
                return a + b;
            }),
            SUB((int a, int b)->{
                System.out.println("The result of " + a + " - " + b + " =");
                return a - b;
            }),
            MUL((int a, int b)->{
                System.out.println("The result of " + a + " * " + b + " =");
                return a * b;
            }),
            DIV((int a, int b)->{
                System.out.println("The result of " + a + " / " + b + " =");
                return a / b;
            });
            
            private final IOperatorStrategy operator;

            // We can define the enum's constructor here
            EnumOperatorStrategy(IOperatorStrategy operator) {
                this.operator = operator;
            }

            @Override
            public int calculate(int a, int b) {
                return operator.calculate(a, b);
            }
        }
    }

    private class Calculator {
        private IOperatorStrategy operator;

        public Calculator(IOperatorStrategy operator) {
            this.operator = operator;
        }

        public int calculate(int a, int b) {
            return operator.calculate(a, b);
        }
    }

    public void testCalculator() {
        Calculator calculator = new Calculator(LambdaStartegy.EnumOperatorStrategy.ADD);
        System.out.println(calculator.calculate(1, 2));

        calculator = new Calculator(new SubOperator());
        System.out.println(calculator.calculate(1, 2));

        calculator = new Calculator(new MulOperator());
        System.out.println(calculator.calculate(1, 2));

        calculator = new Calculator(new DivOperator());
        System.out.println(calculator.calculate(1, 2));
    }


    public static void main(String[] args) {
        Strategy strategy = new Strategy();
        strategy.testCalculator();
    }
}